pkg/signal/main.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package signal
+
+import (
+	"sync"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/utils"
+)
+
+// ErrNilCallback throws if there is no callback function provided
+var ErrNilCallback = errs.Class("must provide a callback function to register")
+
+// Dispatcher is what's loaded onto each service
+type Dispatcher struct {
+	sync.RWMutex
+	source    string
+	callbacks map[string][]Callback
+}
+
+// Callback is the general type for functions in this package
+type Callback func() error
+
+// NewDispatcher is what each service calls to create a dispatch service
+func NewDispatcher(source string) *Dispatcher {
+	return &Dispatcher{
+		source:    source,
+		callbacks: make(map[string][]Callback),
+	}
+}
+
+// Dispatch broadcasts an event across all callbacks
+// If any of them error, it will return that error,
+func (d *Dispatcher) Dispatch(name string) error {
+	d.Lock()
+	defer d.Unlock()
+
+	errors := []error{}
```

Comment:
as soon as utils.ErrorGroup is in, you can use it for this: https://github.com/storj/storj/pull/636/files#diff-fb81b0a96eed5b010cfb081b3aef4431R82